### ISSUE TITLE
Enhance the Climatic > File > CLIMSOFT dialogue

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#4915]
There is a bug that I hope can be fixed by Verison 4.23.  Then there are various enhancements that would be useful in their own right, and would also set a good example for CLIMSOFT dialogues.

Only the bug need be fixed for Rwanda - other stuff would be nice!
1. The bug is that the connection to the database can only be made once in a session.  It works fine.  But if you try again after disconnection, then it gives an error.  It has a line in the error message I don't quite understand which is something like widget followed by a number.  This number gets one larger every time you try the connection.
2. **The sub-dialogue** (making the connection could now be made easier to use.  There are 4 fields.
The first is the name of the database.  There are 2 names now.  I suggest the first is given as the default, and there is a pull-down, tha currently just contains the 2 names.  But there could be more later I assume. The first (default) is mariadb_climsoft_db_v4 and the second is mariadb_climsoft_test_db_v4.
The third field could perhaps also be a pull-down.  It contains 3308 as the default and I'll check with Marcellin what other values might sensibly be there.  For now also have 3306.
3. The **main dialogue** could be made a bit easier.  It could also add flexibility in what is copied over.  Some aspects need more thought, but: 
a) A simple improvement would be to be able to display the station names as well, or instead of the IDs in the selector.  And possibly, with a tooltip, to be able to display other station information, particularly the start (and possibly the end) of the record at that station.
b) A second could be to use the Element information in the dialogue.  Elements each have a **type**.  They are Daily, Monthly, Dekadal, Hourly, Synop and AWS and All.  If just one type is selected, then only the elements of that type are shown.  Eventaully perhaps only those stations are shown in the selector that has data of that type - but that might be tricky with the current software.
What would be useful from that, is the date-time column that is in the output.  It is currently not a date column, but could become one if the data were daily or monthly.
c)  There could also usefully be a checkbox for Transfer Element table.  Currently the Element information is given in the data table.  It is given in 3 ways, and that is tricky, because when made into factors, those columns will be in different orders of their levels. It would be better to just give the element number abd give the other information in the element table.
d) When multiple elements are transferred there could be a Tidy checkbox.  This does an unstack.
d) Later there might also be an Options button perhaps, as there is in the NetCDF dialogue.  
3. a) The main output data frame would become simpler in just giving the Element numbers.  And this column is made into a factor.
b) The Date column is made into a date when the data are daily or monthly.
c) I could not format the data column (perhaps these small numbers again?
d) Perhaps the data could be unstacked, when multiple elements are output.  I am nogt convinced this will always work - for example if there are duplicate values.  So it should be an option.

**Comment by maxwellfundi** [comment]
@dannyparsons Am aware we fixed the bug that was to be fixed on here. Have you got time this week to have a look at other things needed here?

**Comment by dannyparsons** [comment]
I'm looking into some of these things for the release.

**Comment by rdstern** [comment]
The current version reads some files but fails on reading the Rwanda station data.  On inspection this is because some of the stations have non-numeric station ID columns.  Can it just read them in as strings?

I hope this might be simple enough to fix this week?  We are now planning for a quick update on Friday, because of the bug in the new data_book, that made the QC temperature dialogue unusable.

**Comment by dannyparsons** [comment]
@maxwellfundi have you looked into this? I don't see why the database has the station ID column as string but should be able to do something on the R side if necessary. Is it clear why this gives an error?

**Comment by maxwellfundi** [comment]
@dannyparsons  you are wrong on this. We found out that the station ID is can acually be a string. For the Rwanda met they have some stations id that have alphabe letters in them eg 13452PLR as the station Id. 

So if you could change this so that it accepts this kind of id too would be great!

**Comment by rdstern** [comment]
If the input of the stations is to be a text, rather than numeric, then I think all that needs to change is to put quotes round each station when sending the command to R.  

Can someone do that in the AMI team?  Then (to test it) I think the test database could be edited to add non-numeric characters to some of the station IDs there.  I can test if the change is merged.

It would be good to get this change in the forthcoming version and then yell Verdaste.

**Comment by dannyparsons** [comment]
I believe the bug of non numeric IDs is now fixed. @maxwellfundi can you confirm?

**Comment by maxwellfundi** [comment]
I dont  seem to remember if  this was actually fixed and i think if it was then there would be a reference to the PR that fixed it. 
Looking at the dialog, there are a couple of things that needed to be done that still are pending including 

- [ ] Changing the start date and end date to date pickers 

- [ ] Adding the "Unstack Data"  check box like @rdstern  would like.

**Comment by maxwellfundi** [comment]
@Ivanluv  Could you start with the addition of the checkbox and changing the input textboxes to date pickers.

**Comment by rdstern** [comment]
Now you are in the meeting it could be a good time to plan and possibly implement some of the improvements to the Climsoft dialogue.  I attach the first part of the Climsoft importing for the R-Instat climatic guide.  It doesn't make for easy reading.

Here are some points I noticed in using the current dialogue.
1) When I return to the dialogue (having made a connection) the top button says "Establish Connection"  It should know that one was established.
2) The element information is also cleared each time. 
3) The date-time column is there.  Could the importing also produce a date column.  It would save one step.  It is really easy from the Make Date dialogue as it is the default, so I hope will be easy to add.  (We probably need to keep the date-time column, because we can have sub-daily data.
4) Could the **Abbreviation** column be made into a factor.  Could it be called Element?
5) Could the column called **recordedFrom** be called Station?
6) Do we need to transfer all 3 columns of element information?  If just one then the Abbreviation (now I hope called Element) Ideally, as it is a factor, perhaps the element numbers could be the levels?
7) A bigger change would be to have an option in the Climsoft dialogue to do the unstacking - if multiple elements are chosen.  However that's probably more work, so could wait.

These changes would shorten the document a bit.  We want to show that this process is simple.

[R-Instat Climatic Guide 4.5 Climsoft.docx](https://github.com/africanmathsinitiative/R-Instat/files/2975742/R-Instat.Climatic.Guide.4.5.Climsoft.docx)

**Comment by rdstern** [comment]
I think there has been some work on this issue?  Otherwise it can be moved to Milestone 4.33.

**Comment by maxwellfundi** [comment]
@rdstern  yes, the date pickers were added already. Now its the issue of stacking the data and then addressing the issues you have raised above.

**Comment by Patowhiz** [comment]
@rdstern on your original item 2, The sub-dialogue. Do we have a reason as to why the we have a separate R dialog for entering the password. I initially thought that this was the only way to pass the password to R. But digging deeper into the R database driver interface, that's not the case. If there is no concrete reason for the extra R dialog, then in addition your design proposals, I came up with the sketch below. What are your thoughts on it?

![sketch](https://user-images.githubusercontent.com/21179007/91566553-89636400-e94c-11ea-82af-082d60632ba2.jpg)

**Comment by Patowhiz** [comment]
@rdstern. Below is the current state of the dialog. I also have to say, the choice of using date pickers makes it impossible to not select dates. Date pickers don't accept null or empty inputs. Will we always require users to enter date ranges? If not, then I suggest looking into another form of input or introducing another checkbox that toggles between selecting date range or not.

<img width="203" alt="current dialog" src="https://user-images.githubusercontent.com/21179007/91567312-9896e180-e94d-11ea-9646-7c9f5673daf8.PNG">

Below is the R password prompt dialog that I'm proposing to eliminate, I still can't get a concrete reason why we are using it.

<img width="253" alt="password dialog" src="https://user-images.githubusercontent.com/21179007/91567451-ca0fad00-e94d-11ea-8b59-26c303723696.PNG">

**Comment by Patowhiz** [comment]
@rdstern following our discussions. Following your descriptions Below is a sketch of the main dialog, I have omitted the unstack checkbox for now as that could potentially be a separate issue and a bigger change. The establish connection button will open the database connection subdialog(see sketch above). And of course it will change to Connection Established as outlined in the issue comment above.

![current_proposed_design](https://user-images.githubusercontent.com/21179007/91634717-1c5ad780-e9fb-11ea-90a4-bb8280569ed4.jpg)

**Comment by Patowhiz** [comment]
@rdstern following our discussions on the climsoft import "wizard". Below are my rough sketches on some of the design details. I also welcome @lloyddewit comments between the above proposed design and the wizard sketch.

Ideally, this could be 1 dialog showing a dynamic UI to the user in 3 steps (think of a paginated web form). I find this easier and could be used as a basis of simplifying importing from other structured data I think. 

Step 1 -  connecting to the database
Step 2 - importing selecting stations
Step 3 - selecting elements and range of data to import. 

At any given point the user can go back and forth and can generate the script. 
I'm prioritising on just doing @rdstern proposed design but looking forward to thoughts on this "wizard" design. We can also have both(it's now standard practice to have both a wizard and a direct feature in most softwares e.g Excel, Access etc).

![step_1](https://user-images.githubusercontent.com/21179007/91634916-f0d8ec80-e9fc-11ea-9b33-8f8a0dcac634.jpg)

![step_2](https://user-images.githubusercontent.com/21179007/91634940-14039c00-e9fd-11ea-8593-7d11f99e3fbe.jpg)

![step_3](https://user-images.githubusercontent.com/21179007/91634979-57f6a100-e9fd-11ea-8265-420fd41b05e5.jpg)

**Comment by rdstern** [comment]
@Patowhiz the designs look good.  The one point that is available currently - though perhaps not sufficiently obvious is that you can import just the metadata, i.e. information about the stations, without necessarily importing any data.  I like that.  Ideally that could also include element information, i.e. all stations that have temperature data, but I am not sure that information is (yet) easily available in Climsoft?  At this stage we only import what is currently easily available in Climsoft.

**Comment by Patowhiz** [comment]
@rdstern  oh! yes, for the wizard, we can have a skip this step checkbox, or just allow users be able to not to select any element, either way. 
On the issue of the metadata, currently we have a table that stores metadata about all elements in  a station(you can access this information in the station element metadata form). Though I'm not sure if met services are filling in this metadata information.

**Comment by Patowhiz** [comment]
@rdstern in regards to the items that you have proposed in your comments, below is the main dialog and sub dialog current design and functionality. I have opted to satisfy your requests first before looking into the wizard.

<img width="207" alt="subdialog" src="https://user-images.githubusercontent.com/21179007/91924690-10616500-ecdc-11ea-8b19-63ffd18574f7.PNG">

<img width="335" alt="new main dialog" src="https://user-images.githubusercontent.com/21179007/91924677-08092a00-ecdc-11ea-9c85-e95a61243e32.PNG">

**Comment by Patowhiz** [comment]
Further into you thoughts in regards to the results of the dialog,

1. Would you also like the stations column to be converted to factor?

2. I've decided to use the `abbreviation` only column as requested by you, but then I noticed if we wanted to build a relationship key between the elements metadata and observation data, I think we may want the `elementId` column.

3. In regards to the date-time column, I can't see an obvious way of knowing whether the data being imported is daily or hourly, any ideas about this? Or are you suggesting we have an extra column that is simply just a date and still leave the datetime column?

I have fixed the other features and bugs requested except the unstacking.

**Comment by rdstern** [comment]
Great.  
1. Yes Station column could be a factor.
2. Don't need to change too much from current situation with the elements.  Later we will offer an option to unstack and that will name the variables with the abbreviation.  
3. I don't trust Climsoft to not have duplicate observations.  So we shouldn't have unstacking as mandatory.  I think they are unique, but only if you include the hour of recording.
4. Happy to have date-time (as now) and then also the date - which is what is done when we import from Eumetsat.

**Comment by Patowhiz** [comment]
@rdstern I'm not sure if "now" would be a correct name of the datetime, then I'm also keen on not using database keywords in the sql queries that retrieve the data. Below is a screenshot of the dataview with the results of the data from the dialog.

<img width="396" alt="dataview" src="https://user-images.githubusercontent.com/21179007/91991430-8a750680-ed3b-11ea-8a2a-c94a4b697be4.PNG">

**Comment by rdstern** [comment]
I agree.  Do we need obs?  Could be just datetime and date?  
Looks good.

**Comment by Patowhiz** [comment]
@rdstern below are screenshots of the wizard dialog and its step. Its flexible enough to cater all the items raised in this issue.

Connect to database
<img width="439" alt="1" src="https://user-images.githubusercontent.com/21179007/92293479-9c69cb80-ef2c-11ea-9ee7-14e46de7b450.PNG">

Import stations
<img width="439" alt="2" src="https://user-images.githubusercontent.com/21179007/92293482-a25fac80-ef2c-11ea-999f-de7d0c88cb92.PNG">

Import observation data (Optional)
<img width="438" alt="3" src="https://user-images.githubusercontent.com/21179007/92293484-a5f33380-ef2c-11ea-93eb-06558bc98a2f.PNG">

**Comment by Patowhiz** [comment]
To access the wizard I've put into the Import from climsoft main dialog a temporary button through which you can access. But with your approval I can have another quick PR that creates a menu link through which a user can access the wizard dialog separately without launching this main dialog.(I think this will help reduce confusion during testing)

<img width="338" alt="temporary main dialog" src="https://user-images.githubusercontent.com/21179007/92293578-49444880-ef2d-11ea-8d6c-5940d2a4e68a.PNG">

**Comment by mhabimana** [comment]
@Patowhiz - Thanks for improving on this. I am happy with the design, having both options (manual selection and wizard option) is very good and suitable. I can see you are selecting stationId but when it comes to element, your are selecting element abbreviations, I wonder if possible to bring together both (**stationId;stationName**) and (**elementId;elementName**) displays in the selector and receiver since they are related and then have these dispayed in the dataview table as separate columns but one after the other (i.e.: stationId,StationName,elementId,elementName).

**Comment by Patowhiz** [comment]
@mhabimana thank you for going through this. 
@rdstern had interesting reasons as to why abbreviations would be good, and also why having both ids and abbreviations isn't necessary.  Would be nice to hear his thoughts about your suggestions.

Please note you can now also import the elements information. So any extra information not captured in the observation data dataframe can be views in the elements  data frame.

**Comment by mhabimana** [comment]
@Patowhiz -Thanks,I think is Roger prefers to use the stationId,let the elementId be used too  instead of the element abbreviation,assuming the element abbreviation can be added in the dataframe when required as you said.But as I said earlier,it would be great to have both stationid and elementid in both selector and receiver,

**Comment by mhabimana** [comment]
@Patowhiz - I am looking at this and it looks very good. I am suggesting you to add on this dialog another box and label for selecting or including elements because,in real sense(what a user see after clicking the,**Next** Button,is the dialog for elements(i.e.: the next button calls the dialog where to fill elements details :
![image](https://user-images.githubusercontent.com/10340170/92733456-5ffcfc00-f380-11ea-8c73-735d287e2fa7.png)

**Comment by mhabimana** [comment]
@Patowhiz - Good to name the Checkbox, **Include Elements Info**

**Comment by Patowhiz** [comment]
@mhabimana thanks. Will do this.

**Comment by dannyparsons** [comment]
Sorry I am coming to this issue quite late. I have tried to catch up on the discussion.

> Do we have a reason as to why the we have a separate R dialog for entering the password.

This is done to ensure that the password is never stored in the front end and never appears in any R code that is generated. By using the pop up box directly from R, the password does not need to be passed from the front end into R. So security is the reason. Also, for reproducibility the R code can freely be shared with anyone and there is no sharing of passwords. The R code can be run again and the pop up will appear in the same way and will work for anyone who does have the password but it needs to be typed in every time.

**Wizard Format**

The functionality looks nice and wizard idea is interesting. We do need to have a very general discussion on this concept. Currently, we deliberately have all dialogs which produce R code using the same common design and function in a specific way, which is centred around `ucrButtons`. Even any small changes to this, hiding buttons/changing text, needs to be part of the general discussion. 

So before we have such a dialog done in a different way we need to come back to discuss the original reasons why the dialogs are designed in the way they are now, and why they are all the same, and consider the implications on these original reasons of a new design. Although it's only one dialog at the moment, almost any dialog could be redesigned in a wizard format, so we need to discuss if, when and why we would use this format in general in R-Instat. I think this will be useful to make sure we keep the important aspects of the software as we introduce new ideas.

**Comment by Patowhiz** [comment]
@dannyparsons yes. I noticed why the prompt was from Rm I've retained that functionality. 

As to the ucrButtons, I agree that is part of a much wider discussions. I was keen not to touch the control. It's still the one that I'm using in the wizard. We had a brief discussion with @lloyddewit about it.

Once this code is ready and looked at we can have a separate issue discussing how such wizard dialogs would generally be implemented using our current controls.

**Comment by mhabimana** [comment]
@Patowhiz - I suggest you to keep the Import from Climsoft wizard dialog only instead of both options,the wizard option is straight forward and simple.
![image](https://user-images.githubusercontent.com/10340170/97856110-ea246980-1d0c-11eb-81e5-f306b0e2cfca.png)

**Comment by mhabimana** [comment]
> @Patowhiz - I suggest you to keep the **Import from Climsoft wizard dialog** only instead of both options,the wizard option is straight forward and simple.
> ![image](https://user-images.githubusercontent.com/10340170/97856110-ea246980-1d0c-11eb-81e5-f306b0e2cfca.png)

What is your view on my suggestion of keeping the wizard dialog and not both options? does it make any sense to you?

**Comment by mhabimana** [comment]
> > @Patowhiz - I suggest you to keep the **Import from Climsoft wizard dialog** only instead of both options,the wizard option is straight forward and simple.
> > ![image](https://user-images.githubusercontent.com/10340170/97856110-ea246980-1d0c-11eb-81e5-f306b0e2cfca.png)

What is your view on my suggestion of keeping the wizard dialog and not both options? does it make any sense to you?

**Comment by Patowhiz** [comment]
@mhabimana thank you for your feedback.  Yes it makes sense. The wizard was an idea I had in making it easy for climsoft users to import data, I'm happy you happen to find it so. I'm also happy with either options because it gives user's the freedom to choose either options. 

Will be keen to get other's feedback on the same
